### ISSUE TITLE
Unclear error message when using `Result` instead of `Ok`

### ISSUE THREAD

**Issue by joepie91** [rust-lang/rust#43871]
I made the mistake of trying to do `if let Result(foo) = bar` instead of `if let Ok(foo) = bar`, and got a confusing error like the following:

```
error[E0532]: expected tuple struct/variant, found enum `Result`
  --> src/service/prime.rs:18:16
   |
18 |         if let Result(parsed_number) = request.parse::<u64>() {
   |                ^^^^^^ not a tuple struct/variant
```

It took me a while to figure out what was wrong here; the [error index entry](https://doc.rust-lang.org/error-index.html#E0532) for this error code was not very helpful either. Part of the problem is that the error tells me what it *isn't*, but not how what it *is* relates to that. To a beginner, the relationship between an Enum and one of its variants won't be intuitively obvious.

Perhaps the compiler feedback could also be improved here by suggesting `Ok`/`Err` (for `Result`) and `Some`/`None` (for `Option`), as these are common cases where this might occur. A more generic approach would be to simply list expected/valid variants for the specified Enum.

(As an aside, the contributing guidelines claimed that there would be an issue template, but there doesn't seem to be. I'm not sure whether that is intentional or not.)

**Comment by estebank** [comment]
[`check_pat_tuple_struct`] should look at the `def.kind` and if it is an enum, it should suggest using one of the enum's variants:

```
error[E0532]: expected tuple struct/variant, found enum `Result`
  --> src/service/prime.rs:18:16
   |
18 |         if let Result(parsed_number) = request.parse::<u64>() {
   |                ^^^^^^ not a tuple struct/variant
   = note: did you mean to use one of the following variants?
           - `Ok(T)`
           - `Err(E)`
```


[`check_pat_tuple_struct`]: https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc_typeck/check/_match.rs#L635-L640

**Comment by Menschenkindlein** [comment]
Hello, I'd like to take this one.

**Comment by estebank** [comment]
@Menschenkindlein feel free to reach out if you need any help!

**Comment by Menschenkindlein** [comment]
@estebank I could not reproduce the error `E0164`, that is reported by [check_pat_tuple_struct](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc_typeck/check/_match.rs#L635-L640).
However, I found that the error in the topic is a base error, reported by [smart_resolve_path_fragment](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc_resolve/lib.rs#L2444-L2451). I added a case `(Def::Enum(defId), PathSource::TupleStruct)` to the [match](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc_resolve/lib.rs#L2557), but I cannot find how to get `hir::Item` from `DefId` in this context. I hope, it will give me access to the enum's variants.

**Comment by estebank** [comment]
[`self.tcx.hir.get_if_local(def_id)`](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc/hir/map/mod.rs#L511) will give you a [`Node`](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc/hir/map/mod.rs#L45-L67) if this `DefId` is part of the current crate, which has a `NodeItem` variant that holds a `hir::Item` and from there you can get the `vec` of variants you're looking for.

**Comment by Menschenkindlein** [comment]
Could you, please, explain to me what this magic `tcx` field really is? I saw its usage in `check_pat_tuple_struct`, but there was other `self` - `FnCtxt`. Both it and `Resolver` doesn't have this field in their definition.

**Comment by estebank** [comment]
@Menschenkindlein you have [the following code](https://github.com/rust-lang/rust/blob/master/src/librustc/ty/context.rs#L807-L812):

```rust
impl<'a, 'gcx, 'tcx> Deref for TyCtxt<'a, 'gcx, 'tcx> {
    type Target = &'a GlobalCtxt<'gcx>;
    fn deref(&self) -> &Self::Target {
        &self.gcx
    }
}
```

As `TyCtxt` implements `Deref` for `GlobalCtxt`, when accessing a field of a borrow of `tcx` you can also access the methods and fields of `tcx.gcx`. You can think of it as a way of having [inheritance (with caveats)](https://github.com/rust-unofficial/patterns/blob/master/anti_patterns/deref.md) in rust.

**Comment by Menschenkindlein** [comment]
`rustc_resolver::Resolver` doesn't have access to `GlobalCtxt` as far as I see. I think the change should be made in some of the calling functions. It is too difficult for me to find in which one, though.

**Comment by estebank** [comment]
Ah! Now I see what you mean regarding `Resolver`. So if this is being raised in `librustc_resolver` you don't have access to `tcx` there. I'm gonna have to look how this could be accomplished, but in the meantime maybe [this suggestion code](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc_resolve/lib.rs#L2500-L2521) that uses [`lookup_import_candidates`](https://github.com/rust-lang/rust/blob/17f56c549c35bb2cb316e5abff116e65277c7bb1/src/librustc_resolve/lib.rs#L3399-L3404) could be used as a starting point.

**Comment by Menschenkindlein** [comment]
I implemented [`lookup_enum_variants`](https://github.com/Menschenkindlein/rust/commit/eb8a9c65f2fc7a1e5142c3590cfb4b944384a958#diff-aeb0880081a991f34aef2ab889e1fb7aR3476). It is mostly a copy of `lookup_import_candidates`. It looks like it works:
```
error[E0532]: expected tuple struct/variant, found enum `Result`
  --> src/lib.rs:16:12
   |
16 |     if let Result(y) = x {
   |            ^^^^^^
   |
   = note: did you mean to use one of the following variants?
           - `std::result::Result::Ok`
           - `std::result::Result::Err`
```
However, I don't like both the reuse (misuse) of `ImportSuggestion`, and all that code duplication. Could you, please, suggest, where to move further.

**Comment by estebank** [comment]
Great work, @Menschenkindlein! The code seems good to me. I would probably create a new method being called by the `lookup_*` methods that holds the related code and accepts a `fn` to operate over the set. If I had some extra time, I would also rework it so that it was an iterator, but that might be more effort than is worth, but feel free to attempt it if you want :)

As per the `ImportSuggestion`, we can get away with just renaming it to `PathSuggestion`, as there's nothing intrinsic about that struct that is about imports, other than its current usage.

**Comment by Menschenkindlein** [comment]
It didn't look as a feasible task for me to combine the `lookup_*` functions, because the closure in the common function would have to take a very complex set of arguments. And I would need to put too many constraints into that function. For example, in the search for enum, I don't need the order of modules to be stable, but I would need to use `for_each_child_stable`. That's why I tried to make my function not so similar to `lookup_import_candidates` - I extracted [find_module](https://github.com/Menschenkindlein/rust/blob/4d133f37a8a694ee8fec659d4c998024884ea7e8/src/librustc_resolve/lib.rs#L3480).
Now I have a question, whether `find_module` should return `Option` or `panic!` when the module is not there, or if the passed `def` is not the def of a module. For the latter, I think, it would be better to `panic!`, but I'm not sure if there is an easy way to determine, whether `def` defines `module`.
